# Trees, Forests, Support, SVMs & k-NN

#### Week 3 - Day 4

In this notebook, we will cover:

1. Train a **decision tree**, **random forest**, **SVM**, and **k-NN** on the same train/test split from Day 3.
2. Evaluate all four with the same metric (e.g. **F1-score**) and assemble the results into one comparison table.
3. Report the random forest's top feature importances and interpret them.
4. Identify the best-performing model for this dataset and explain, in Markdown, why it likely won.

> Note: Words written in bold represents key topics that were studied and covered during the day.

## Step 0 - Setup data & Load the dataset:

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (f1_score, confusion_matrix, classification_report, accuracy_score, precision_score, recall_score, roc_auc_score)
import warnings
warnings.filterwarnings('ignore')

sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

In [7]:
from sklearn.datasets import load_iris  # Load "iris" dataset as example
iris = load_iris()
X = pd.DataFrame(iris.data, columns=iris.feature_names)
y = pd.Series(iris.target, name='target')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train) 
X_test_scaled = scaler.transform(X_test)        

print(f"   X_train: {X_train.shape}")
print(f"   X_test: {X_test.shape}")
print(f"Training target distribution:\n{pd.Series(y_train).value_counts().to_dict()}")

   X_train: (120, 4)
   X_test: (30, 4)
Training target distribution:
{0: 40, 2: 40, 1: 40}


## Step 1 - Train a decision tree, random forest, SVM, and k-NN:

In [9]:
# Train a Decision Tree:

dt_model = DecisionTreeClassifier(max_depth=5, random_state=42, min_samples_split=10)
dt_model.fit(X_train, y_train)  
dt_predictions = dt_model.predict(X_test)
dt_f1 = f1_score(y_test, dt_predictions, average='weighted')
dt_accuracy = accuracy_score(y_test, dt_predictions)
dt_precision = precision_score(y_test, dt_predictions, average='weighted')
dt_recall = recall_score(y_test, dt_predictions, average='weighted')

print(f"\nDecision Tree Performance Metrics:")
print(f"   F1-Score (weighted): {dt_f1:.4f}")
print(f"   Accuracy: {dt_accuracy:.4f}")
print(f"   Precision (weighted): {dt_precision:.4f}")
print(f"   Recall (weighted): {dt_recall:.4f}")

# Additional insights:
print(f"\n Tree Information:")
print(f"   Tree depth: {dt_model.get_depth()}")
print(f"   Number of leaves: {dt_model.get_n_leaves()}")
print(f"   Number of features used: {dt_model.n_features_in_}")


Decision Tree Performance Metrics:
   F1-Score (weighted): 0.9666
   Accuracy: 0.9667
   Precision (weighted): 0.9697
   Recall (weighted): 0.9667

 Tree Information:
   Tree depth: 3
   Number of leaves: 5
   Number of features used: 4


In [12]:
# Train Random Forest:

rf_model = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf_model.fit(X_train, y_train)  
rf_predictions = rf_model.predict(X_test)
rf_f1 = f1_score(y_test, rf_predictions, average='weighted')
rf_accuracy = accuracy_score(y_test, rf_predictions)
rf_precision = precision_score(y_test, rf_predictions, average='weighted')
rf_recall = recall_score(y_test, rf_predictions, average='weighted')

print(f"\n Random Forest Performance Metrics:")
print(f"   F1-Score (weighted): {rf_f1:.4f}")
print(f"   Accuracy: {rf_accuracy:.4f}")
print(f"   Precision (weighted): {rf_precision:.4f}")
print(f"   Recall (weighted): {rf_recall:.4f}")


 Random Forest Performance Metrics:
   F1-Score (weighted): 0.8997
   Accuracy: 0.9000
   Precision (weighted): 0.9024
   Recall (weighted): 0.9000


In [14]:
# Train Support Vector Machine (SVM):

svm_model = SVC(kernel='rbf', probability=True, random_state=42, C=1.0)
svm_model.fit(X_train_scaled, y_train)
svm_predictions = svm_model.predict(X_test_scaled)
svm_f1 = f1_score(y_test, svm_predictions, average='weighted')
svm_accuracy = accuracy_score(y_test, svm_predictions)
svm_precision = precision_score(y_test, svm_predictions, average='weighted')
svm_recall = recall_score(y_test, svm_predictions, average='weighted')

print(f"\n SVM Performance Metrics:")
print(f"   F1-Score (weighted): {svm_f1:.4f}")
print(f"   Accuracy: {svm_accuracy:.4f}")
print(f"   Precision (weighted): {svm_precision:.4f}")
print(f"   Recall (weighted): {svm_recall:.4f}")

print(f"\n SVM Information:")
print(f"   Number of support vectors: {len(svm_model.support_vectors_)}")
print(f"   Total training samples: {X_train.shape[0]}")
print(f"   Support vector ratio: {len(svm_model.support_vectors_)/X_train.shape[0]:.2%}")

print(f"\n Interpretation:")
print(f"   Support vectors are the 'important' points that define the decision boundary")
print(f"   Lower ratio = more efficient model")
print(f"   Higher ratio = more complex decision boundary needed")


 SVM Performance Metrics:
   F1-Score (weighted): 0.9666
   Accuracy: 0.9667
   Precision (weighted): 0.9697
   Recall (weighted): 0.9667

 SVM Information:
   Number of support vectors: 47
   Total training samples: 120
   Support vector ratio: 39.17%

 Interpretation:
   Support vectors are the 'important' points that define the decision boundary
   Lower ratio = more efficient model
   Higher ratio = more complex decision boundary needed


In [15]:
# Train K-Nearest Neighbors (K-NN):

knn_model = KNeighborsClassifier(n_neighbors=5, metric='euclidean')
knn_model.fit(X_train_scaled, y_train)
knn_predictions = knn_model.predict(X_test_scaled)
knn_f1 = f1_score(y_test, knn_predictions, average='weighted')
knn_accuracy = accuracy_score(y_test, knn_predictions)
knn_precision = precision_score(y_test, knn_predictions, average='weighted')
knn_recall = recall_score(y_test, knn_predictions, average='weighted')

print(f"\n k-NN Performance Metrics:")
print(f"   F1-Score (weighted): {knn_f1:.4f}")
print(f"   Accuracy: {knn_accuracy:.4f}")
print(f"   Precision (weighted): {knn_precision:.4f}")
print(f"   Recall (weighted): {knn_recall:.4f}")

print(f"\n k-NN Information:")
print(f"   k value (neighbors): {knn_model.n_neighbors}")
print(f"   Training samples stored: {X_train.shape[0]}")
print(f"   Metric: {knn_model.metric}")

print(f"\n Complexity Analysis:")
print(f"   Training time: O(1) - just stores data")
print(f"   Prediction time: O(N*D) per sample")
print(f"   Where N = training samples ({X_train.shape[0]}), D = dimensions ({X_train.shape[1]})")

print(f"\n Interpretation:")
print(f"   Every prediction requires calculating distance to {X_train.shape[0]} points")
print(f"   This is why k-NN is slow on large datasets!")


 k-NN Performance Metrics:
   F1-Score (weighted): 0.9327
   Accuracy: 0.9333
   Precision (weighted): 0.9444
   Recall (weighted): 0.9333

 k-NN Information:
   k value (neighbors): 5
   Training samples stored: 120
   Metric: euclidean

 Complexity Analysis:
   Training time: O(1) - just stores data
   Prediction time: O(N*D) per sample
   Where N = training samples (120), D = dimensions (4)

 Interpretation:
   Every prediction requires calculating distance to 120 points
   This is why k-NN is slow on large datasets!


## Step 2+3+4 - Create Comprehensive Comparison Table & Identify the best-performing model:

In [23]:
comparison_df = pd.DataFrame({
    'Model': ['Decision Tree', 'Random Forest', 'SVM', 'k-NN'],
    'F1-Score': [dt_f1, rf_f1, svm_f1, knn_f1],
    'Accuracy': [dt_accuracy, rf_accuracy, svm_accuracy, knn_accuracy],
    'Precision': [dt_precision, rf_precision, svm_precision, knn_precision],
    'Recall': [dt_recall, rf_recall, svm_recall, knn_recall]
})

comparison_df = comparison_df.sort_values('F1-Score', ascending=False).reset_index(drop=True)

print("\n MODEL PERFORMANCE COMPARISON:")
print("\nRanked by F1-Score (higher is better):\n")
print(comparison_df.to_string(index=False))

# Find the winner
winner_name = comparison_df.iloc[0]['Model']
winner_f1 = comparison_df.iloc[0]['F1-Score']
runner_up_f1 = comparison_df.iloc[1]['F1-Score']
difference = winner_f1 - runner_up_f1

print(f"\n WINNER: {winner_name}")
print(f"   F1-Score: {winner_f1:.4f}")
print(f"   Advantage over 2nd place: {difference:.4f} ({difference*100:.2f}%)")

# Detailed ranking
print(f"\n Detailed Rankings:")
medals = ["A.", "B.", "C.", "D."]
for idx, row in comparison_df.iterrows():
    print(f"   {medals[idx]} {row['Model']:<18} | F1: {row['F1-Score']:.4f} | Acc: {row['Accuracy']:.4f}")


best_predictions = comparison_df.iloc[0]['Model']
if best_predictions == 'Decision Tree':
    best_pred_array = dt_predictions
    best_model_obj = dt_model
elif best_predictions == 'Random Forest':
    best_pred_array = rf_predictions
    best_model_obj = rf_model
elif best_predictions == 'SVM':
    best_pred_array = svm_predictions
    best_model_obj = svm_model
else:
    best_pred_array = knn_predictions
    best_model_obj = knn_model

# Confusion Matrix:
cm = confusion_matrix(y_test, best_pred_array)
print(f"\n Confusion Matrix:")
print(cm)

# Classification Report:
print(f"\n Detailed Classification Report:")
print(classification_report(y_test, best_pred_array, target_names=iris.target_names))


 MODEL PERFORMANCE COMPARISON:

Ranked by F1-Score (higher is better):

        Model  F1-Score  Accuracy  Precision   Recall
Decision Tree  0.966583  0.966667   0.969697 0.966667
          SVM  0.966583  0.966667   0.969697 0.966667
         k-NN  0.932660  0.933333   0.944444 0.933333
Random Forest  0.899749  0.900000   0.902357 0.900000

 WINNER: Decision Tree
   F1-Score: 0.9666
   Advantage over 2nd place: 0.0000 (0.00%)

 Detailed Rankings:
   A. Decision Tree      | F1: 0.9666 | Acc: 0.9667
   B. SVM                | F1: 0.9666 | Acc: 0.9667
   C. k-NN               | F1: 0.9327 | Acc: 0.9333
   D. Random Forest      | F1: 0.8997 | Acc: 0.9000

 Confusion Matrix:
[[10  0  0]
 [ 0  9  1]
 [ 0  0 10]]

 Detailed Classification Report:
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      0.90      0.95        10
   virginica       0.91      1.00      0.95        10

    accuracy                   

## Why is the Decision Tree the Best Model?

The **Decision Tree** achieved the highest **F1-Score (0.9666)**, making it the best-performing model for this classification task.

Although the **SVM** achieved the exact same F1-Score and Accuracy, the Decision Tree was ranked first because it appeared first after sorting the comparison table. In practice, both models performed equally well on this dataset.

### Why the Decision Tree Performed Well:
- It captured the relationships between the features and the target effectively.
- It produced an excellent balance between **Precision (0.9697)** and **Recall (0.9667)**, resulting in a high F1-Score.
- It achieved a high **Accuracy (96.67%)**, correctly classifying most test samples.
- Decision Trees are also highly interpretable, allowing us to visualize and understand the decision rules.

### Conclusion:
Based on the evaluation metrics, the **Decision Tree** was selected as the best model because it achieved the highest overall classification performance while maintaining an excellent balance between precision and recall. However, since the SVM obtained identical performance metrics, either model would be a strong choice for this dataset.

> **Note: Every cell of above is explaied by markdown cell or by comments.**